### PR TITLE
Speed up recover

### PR DESCRIPTION
**Pull request by HTHou**
## Description

This PR introduced a new type of file to speed up the recovery. 
https://timechor.feishu.cn/docx/V5EedcSvjof5PaxqD01cuVIcnRe

## Test
conf/confignode-env.sh MAX_HEAP_SIZE="8G"
conf/datanode-env.sh MAX_HEAP_SIZE="16G"
MAX_DIRECT_MEMORY_SIZE="32G"

conf/iotdb-system.properties 
series_slot_num=10000
schema_region_group_extension_policy=CUSTOM
default_schema_region_group_num_per_database=10
data_region_group_extension_policy=CUSTOM
default_data_region_group_num_per_database=50
time_partition_interval=86400000
enable_seq_space_compaction=false
enable_unseq_space_compaction=false
enable_cross_space_compaction=false
config_node_ratis_snapshot_trigger_threshold=10000

Benchmark config:
```
LOOP=10
DEVICE_NUMBER=10000
REAL_INSERT_RATE=1.0
SENSOR_NUMBER=5
IS_SENSOR_TS_ALIGNMENT=true
TS_ALIGNMENT_RATIO=1.0
IS_CLIENT_BIND=true
CLIENT_NUMBER=100
GROUP_NAME_PREFIX=g_
DEVICE_NAME_PREFIX=d1_
SENSOR_NAME_PREFIX=s_
GROUP_NUMBER=1
IOTDB_SESSION_POOL_SIZE=50
TEMPLATE=false
VECTOR=false
STRING_LENGTH=8
DOUBLE_LENGTH=2
INSERT_DATATYPE_PROPORTION=0:1:1:1:1:1
OPERATION_PROPORTION=1:0:0:0:0:0:0:0:0:0:0
BATCH_SIZE_PER_WRITE=365
CREATE_SCHEMA=true
START_TIME=2012-8-30T08:00:00+08:00
```

On a 16C 32G ubuntu. 
Before this PR, the recovery costs **140s**. After this pr, it costs **8s**.

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      boolean result =
+          taskQueue.offer(
+              () -> {
+                try {
+                  ByteBuffer buffer = ByteBuffer.allocate(5 * Long.BYTES * tsFileResources.size());
+                  for (TsFileResource tsFileResource : tsFileResources) {
+                    tsFileResource.serializeFileTimeIndexToByteBuffer(buffer);
+                  }
+                  buffer.flip();
+                  writer.write(buffer);
```
Comment (jt2594838) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:97]: Better to add some comments for the 5 * 8 bytes.
It repeats several times, may use a static variable.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                  LOGGER.warn("Meet error when record FileTimeIndexCache: {}", e.getMessage());
+                }
+              });
+      if (!result) {
+        LOGGER.warn("Meet error when record FileTimeIndexCache");
+      }
+    }
+  }
+
+  public void logFileTimeIndex(TsFileResource tsFileResource) {
```
Comment (jt2594838) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:108]: The two methods are highly similar. How about using `public void logFileTimeIndex(TsFileResource... tsFileResources)`

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    FileTimeIndexCacheWriter writer =
+        getWriter(
+            dataRegionId,
+            partitionId,
+            StorageEngine.getDataRegionSystemDir(dataBaseName, String.valueOf(dataRegionId)));
+
+    int currentResourceCount =
+        (sequenceFiles == null ? 0 : sequenceFiles.size())
+            + (unsequenceFiles == null ? 0 : unsequenceFiles.size());
+    if (writer.getLogFile().length() > currentResourceCount * (5 * Long.BYTES) * 100) {
```
Comment (jt2594838) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:149]: Should add some comments and give the `100` a proper name.

Comment (jt2594838) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:149]: After the task is submitted and before it is executed, the condition is highly likely to remain true.
Is it possible that several redundant tasks are submitted during this period?

**Review by jt2594838 — APPROVED** [review]

File: integration-test/src/main/java/org/apache/iotdb/it/env/EnvType.java [hunk 4]
```diff
@@ -28,7 +28,7 @@ public enum EnvType {
   TABLE_CLUSTER1;
 
   public static EnvType getSystemEnvType() {
-    String envValue = System.getProperty("TestEnv", Simple.name());
+    String envValue = System.getProperty("TestEnv", Cluster1.name());
     return EnvType.valueOf(envValue);
```
[the remark this answers is not in the document]
Comment (HTHou) [reply integration-test/src/main/java/org/apache/iotdb/it/env/EnvType.java:32]: It should be a typo when resolve the conflict, let me revert the change.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/StorageEngine.java [hunk 5]
```diff
@@ -178,17 +180,18 @@ public static void blockInsertionIfReject() throws WriteProcessRejectException {
...
-    return isAllSgReady.get();
+  public boolean isReadyForReadAndWrite() {
+    return isReadyForReadAndWrite.get();
   }
 
-  public void setAllSgReady(boolean allSgReady) {
-    isAllSgReady.set(allSgReady);
+  public boolean isReady() {
+    return isReady.get();
   }
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/StorageEngine.java:189]: OK, let me rename `isReady` to `isReadyForNonReadWriteFunctions`.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java (hunk 1 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:97]: makes sense

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java [hunk 6]
```diff
@@ -607,16 +616,26 @@ private void recover() throws DataRegionException {
       throw new DataRegionException(e);
     }
 
-    initCompactionSchedule();
+    if (asyncTsFileResourceRecoverTaskList.isEmpty()) {
+      initCompactionSchedule();
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java:620]: Yes, it should be called after all tasks of a region finish. This part is missing. Let me add it.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/HashLastFlushTimeMap.java [hunk 7]
```diff
@@ -94,10 +79,52 @@ public void updateMultiDeviceFlushedTime(
...
+        memIncr += HASHMAP_NODE_BASIC_SIZE + entry.getKey().ramBytesUsed();
+        newDeviceLastFlushTime.updateLastFlushTime(entry.getKey(), entry.getValue());
+        maxFlushTime = Math.max(maxFlushTime, entry.getValue());
+      }
+      long finalMemIncr = memIncr;
+      memCostForEachPartition.compute(
+          timePartitionId, (k1, v1) -> v1 == null ? finalMemIncr : v1 + finalMemIncr);
+    } else {
+      // should not go here
+      long memIncr = 0;
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/HashLastFlushTimeMap.java:104]: Sure.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java (hunk 2 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/tsfile/timeindex/FileTimeIndexCacheRecorder.java:108]: makes sense

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/HashLastFlushTimeMap.java [hunk 8]
```diff
@@ -94,10 +79,52 @@ public void updateMultiDeviceFlushedTime(
...
-        path, (k, v) -> v == null ? time : Math.max(v, time));
+  public void updatePartitionFlushedTime(long timePartitionId, long maxFlushedTime) {
+    ILastFlushTime flushTimeMapForPartition =
+        partitionLatestFlushedTime.computeIfAbsent(
+            timePartitionId, id -> new PartitionLastFlushTime(maxFlushedTime));
+
+    // todo
+    long memIncr = Long.BYTES;
+    flushTimeMapForPartition.updateLastFlushTime(null, maxFlushedTime);
+    memCostForEachPartition.putIfAbsent(timePartitionId, memIncr);
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/HashLastFlushTimeMap.java]: There is some missing code, added.

**Review by jt2594838 — COMMENTED** [review]
LGTM

**Review by jt2594838 — APPROVED** [review]

**Pull request merged**